### PR TITLE
Build VS installers for WebAssembly and Mobile workloads

### PR DESCRIPTION
**Pull request by directhex**
This is a continuation of work by @bekir-ozturk in https://dev.azure.com/dnceng/internal/_git/dotnet-runtime/pullrequest/16294

Last private run (since it touches runtime-official.yml) at https://dev.azure.com/dnceng/internal/_build/results?buildId=1240576&view=results - it's still failing in Icon.png, which is supposed to be fixed by https://github.com/dotnet/arcade/pull/7614 but I'm having issues getting the correct version picked up locally whilst the wholesale update from maestro is still up in the air. Will keep iterating on it, but wanted to open up to wider review.

### CODE REVIEW

**Comment by directhex** [comment]
Official build, for whoever wants to validate the output https://dev.azure.com/dnceng/internal/_build/results?buildId=1242483&view=results

File: src/workloads/workloads.csproj [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+                          Description="Build tools for Android compilation and native linking."/>
+      <ComponentResources Include="microsoft-net-runtime-android-aot" Title=".NET Android Build Tools (AoT)"
+                          Description="Build tools for Android ahead-of-time (AoT) compilation and native linking."/>
+
+      <!-- Visual Studio components must be versioned. Build tasks will fall back to cobbling a version number from
+           the manifest information unless it's overridden. -->
+      <ComponentVersions Include="wasm-tools" Version="$(FileVersion)" />
+      <ComponentVersions Include="microsoft-net-runtime-android" Version="$(FileVersion)" />
+      <ComponentVersions Include="microsoft-net-runtime-android-aot" Version="$(FileVersion)" />
+    </ItemGroup>
```
Comment (pjcollins) [inline src/workloads/workloads.csproj:66]: To save a few lines the `%(Version)` metadata can be added to the corresponding `@(ComponentResources)` above, for example:

```xml
<ComponentResources Include="microsoft-net-runtime-android" Version="$(FileVersion)" Title=".NET Android Build Tools"
                    Description="Build tools for Android compilation and native linking."/>
```

This item can be then passed to both task inputs:

```xml
<GenerateVisualStudioWorkload IntermediateBaseOutputPath="$(WorkloadIntermediateOutputPath)"
                              ComponentResources="@(ComponentResources)"
                              ComponentVersions="@(ComponentResources)"
...
```

File: src/workloads/workloads.csproj [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    <ItemGroup>
+      <!-- Overrides for Visual Studio setup generation. If the workload definition IDs change,
+           these must be updated. -->
+      <ComponentResources Include="wasm-tools" Title=".NET WebAssembly Build Tools"
+                          Description="Build tools for WebAssembly ahead-of-time (AoT) compilation and native linking."/>
+      <ComponentResources Include="microsoft-net-runtime-android" Title=".NET Android Build Tools"
+                          Description="Build tools for Android compilation and native linking."/>
+      <ComponentResources Include="microsoft-net-runtime-android-aot" Title=".NET Android Build Tools (AoT)"
+                          Description="Build tools for Android ahead-of-time (AoT) compilation and native linking."/>
+
```
Comment (pjcollins) [inline src/workloads/workloads.csproj:60]: Could we add `microsoft-net-runtime-ios` and `microsoft-net-runtime-maccatalyst` to this PR as well? I think that would fulfill all of the current xamarin/MAUI requirements.

Comment (directhex) [reply src/workloads/workloads.csproj:60]:
```
  microsoft-net-runtime-ios platforms does not support Windows and will be skipped (osx-arm64, osx-x64).
  microsoft-net-runtime-maccatalyst platforms does not support Windows and will be skipped (osx-arm64, osx-x64).
  microsoft-net-runtime-macos platforms does not support Windows and will be skipped (osx-arm64, osx-x64).
  microsoft-net-runtime-tvos platforms does not support Windows and will be skipped (osx-arm64, osx-x64).
  ```

No.

Comment (lewing) [reply src/workloads/workloads.csproj:60]: what packages does windows need for them?

Comment (steveisok) [reply src/workloads/workloads.csproj]: For ios/tvos/catalyst, we will need to have the runtime packs be in a separate workload and installable on windows.

I am leaning towards having that work done in a quick follow up.

Comment (pjcollins) [reply src/workloads/workloads.csproj]: We are currently trying to include the iOS and MacCatalyst (and perhaps eventually tvOS and macOS) workloads from xamarin-macios in VS on windows, and those extend `microsoft-net-runtime-ios` and `microsoft-net-runtime-maccatalyst`. The manifest generated from the iOS and MacCatalyst workloads currently lists these runtime workloads as dependencies, and VS won't build without them. Plans to try to support this in a separate PR sound good to me.

Comment (steveisok) [reply src/workloads/workloads.csproj]: https://github.com/dotnet/runtime/pull/55851

**Comment by directhex** [comment]
Current versioning story is, on an Official Build, the directory the JSON gets shoved in looks official-ish (6.0.21.27502 from my usual example OfficialBuildId of 20210525.2). Inspecting an MSI with Orca has a ProductVersion of 6.0.0 and a PackageVersion of 6.0.0-preview.7.21365.21.

**Comment by joeloff** [comment]
> Current versioning story is, on an Official Build, the directory the JSON gets shoved in looks official-ish (6.0.21.27502 from my usual example OfficialBuildId of 20210525.2). Inspecting an MSI with Orca has a ProductVersion of 6.0.0 and a PackageVersion of 6.0.0-preview.7.21365.21.

That's good for pack installers. We should check the generated VS authoring next.

**Comment by directhex** [comment]
Latest private build, with version numbered artifacts to check, is https://dev.azure.com/dnceng/internal/_build/results?buildId=1248191&view=results

wixpack.zip are in the `IntermediateArtifacts/workloads/` folder, jsons and MSIs in `Insertion/`

File: src/workloads/VSSetup.props [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        <DropParamService>-s &quot;$(DropServiceUri)&quot;</DropParamService>
+        <DropParamTimeout>--timeout &quot;$(DropTimeout)&quot;</DropParamTimeout>
+        <DropParamTraceLevel>--tracelevel &quot;$(DropTraceLevel)&quot;</DropParamTraceLevel>
+        <DropParamExpirationDate>-x &quot;$(DropExpirationDate)&quot;</DropParamExpirationDate>
+        <!-- Use AAD for authentication -->
+        <DropParamAuth>-a</DropParamAuth>
+    </PropertyGroup>
+
+    <PropertyGroup>
+        <ManifestTeamProject Condition="'$(ManifestTeamProject)' == ''">dotnet</ManifestTeamProject>
```
Comment (pjcollins) [inline src/workloads/VSSetup.props:29]: I think we should add a couple of props here to control the name of the top level .vsman file that is generated, something like:

```xml
    <TargetName>Microsoft.NET.Workload.Runtime</TargetName>
    <ManifestName>$(TargetName)</ManifestName>
```

Comment (pjcollins) [reply src/workloads/VSSetup.props:29]: The file vsman rename will work as well, feel free to mark this as resolved.

**Review by joeloff — APPROVED** [review]
Looks good on the workload side.  I'll defer the YAML changes to someone else who understands the runtime build better

File: eng/Subsets.props [hunk 4]
```diff
@@ -109,7 +109,8 @@
     <SubsetName Include="Mono.Packages" Description="The projects that produce NuGet packages for the Mono runtime." />
     <SubsetName Include="Mono.WasmRuntime" Description="The WebAssembly runtime." />
     <SubsetName Include="Mono.MsCorDbi" Description="The implementation of ICorDebug interface." />
-
+    <SubsetName Include="Mono.Workloads" OnDemand="true" Description="Builds the installers and the insertion metadata for Blazor workloads." />
```
Comment (steveisok) [inline eng/Subsets.props:112]: Description needs updated in a follow up.

**Review by steveisok — APPROVED** [review]

**Pull request merged**